File: app/src/main/java/biz/dealnote/messenger/fragment/PlaylistFragment.java
package biz.dealnote.messenger.fragment;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.appcompat.app.ActionBar;
import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.google.android.material.floatingactionbutton.FloatingActionButton;

import java.util.ArrayList;

import biz.dealnote.messenger.Extra;
import biz.dealnote.messenger.R;
import biz.dealnote.messenger.activity.ActivityFeatures;
import biz.dealnote.messenger.activity.ActivityUtils;
import biz.dealnote.messenger.adapter.AudioRecyclerAdapter;
import biz.dealnote.messenger.fragment.base.BaseFragment;
import biz.dealnote.messenger.listener.BackPressCallback;
import biz.dealnote.messenger.model.Audio;
import biz.dealnote.messenger.place.Place;
import biz.dealnote.messenger.place.PlaceFactory;
import biz.dealnote.messenger.player.MusicPlaybackService;
import biz.dealnote.messenger.player.util.MusicUtils;
import biz.dealnote.messenger.settings.Settings;
import biz.dealnote.messenger.util.PhoenixToast;

import static biz.dealnote.messenger.util.Objects.isNull;

/**
 * Created by golde on 27.09.2016.
 */
public class PlaylistFragment extends BaseFragment implements AudioRecyclerAdapter.ClickListener,
        BackPressCallback {

    private RecyclerView mRecyclerView;
    private View root;
    private AudioRecyclerAdapter mAdapter;
    private ArrayList<Audio> mData;
    private PlaybackStatus mPlaybackStatus;

    public static Bundle buildArgs(ArrayList<Audio> playlist) {
        Bundle bundle = new Bundle();
        bundle.putParcelableArrayList(Extra.AUDIOS, playlist);
        return bundle;
    }

    public static PlaylistFragment newInstance(ArrayList<Audio> playlist) {
        return newInstance(buildArgs(playlist));
    }

    public static PlaylistFragment newInstance(Bundle args) {
        PlaylistFragment fragment = new PlaylistFragment();
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (savedInstanceState != null) {
            restoreFromSavedInstanceState(savedInstanceState);
        }
        mData = requireArguments().getParcelableArrayList(Extra.AUDIOS);
    }

    private int getAudioPos(Audio audio) {
        if (mData != null && !mData.isEmpty()) {
            int pos = 0;
            for (final Audio i : mData) {
                if (i.getId() == audio.getId() && i.getOwnerId() == audio.getOwnerId()) {
                    i.setAnimationNow(true);
                    mAdapter.notifyDataSetChanged();
                    return pos;
                }
                pos++;
            }
        }
        return -1;
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        root = inflater.inflate(R.layout.fragment_playlist, container, false);
        ((AppCompatActivity) requireActivity()).setSupportActionBar(root.findViewById(R.id.toolbar));

        mRecyclerView = root.findViewById(R.id.list);
        LinearLayoutManager manager = new LinearLayoutManager(requireActivity(), RecyclerView.VERTICAL, false);
        mRecyclerView.setLayoutManager(manager);
        FloatingActionButton Goto = root.findViewById(R.id.goto_button);

        Goto.setOnLongClickListener(v -> {
            Audio curr = MusicUtils.getCurrentAudio();
            if (curr != null) {
                PlaceFactory.getPlayerPlace(Settings.get().accounts().getCurrent()).tryOpenWith(requireActivity());
            } else
                PhoenixToast.CreatePhoenixToast(requireActivity()).showToastError(R.string.null_audio);
            return false;
        });
        Goto.setOnClickListener(v -> {
            Audio curr = MusicUtils.getCurrentAudio();
            if (curr != null) {
                int index = getAudioPos(curr);
                if (index >= 0) {
                    if (Settings.get().other().isShow_audio_cover())
                        mRecyclerView.scrollToPosition(index);
                    else
                        mRecyclerView.smoothScrollToPosition(index);
                } else
                    PhoenixToast.CreatePhoenixToast(requireActivity()).showToast(R.string.audio_not_found);
            } else
                PhoenixToast.CreatePhoenixToast(requireActivity()).showToastError(R.string.null_audio);
        });
        return root;
    }

    @Override
    public void onViewCreated(View view, Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        mAdapter = new AudioRecyclerAdapter(requireActivity(), mData, false, false);
        mAdapter.setClickListener(this);
        mRecyclerView.setAdapter(mAdapter);
    }

    private void restoreFromSavedInstanceState(@NonNull Bundle state) {
        //this.mData = state.getParcelableArrayList(Extra.AUDIOS);
    }

    @Override
    public void onClick(int position, Audio audio) {
        MusicPlaybackService.startForPlayList(requireActivity(), mData, position, false);
        if (!Settings.get().other().isShow_mini_player())
            PlaceFactory.getPlayerPlace(Settings.get().accounts().getCurrent()).tryOpenWith(requireActivity());
    }

    @Override
    public void onResume() {
        super.onResume();
        Settings.get().ui().notifyPlaceResumed(Place.AUDIO_CURRENT_PLAYLIST);

        ActionBar actionBar = ActivityUtils.supportToolbarFor(this);
        if (actionBar != null) {
            actionBar.setTitle(R.string.playlist);
            actionBar.setSubtitle(null);
        }
        new ActivityFeatures.Builder()
                .begin()
                .setHideNavigationMenu(false)
                .setBarsColored(requireActivity(), true)
                .build()
                .apply(requireActivity());
        this.mPlaybackStatus = new PlaybackStatus();
        final IntentFilter filter = new IntentFilter();
        filter.addAction(MusicPlaybackService.PLAYSTATE_CHANGED);
        filter.addAction(MusicPlaybackService.SHUFFLEMODE_CHANGED);
        filter.addAction(MusicPlaybackService.REPEATMODE_CHANGED);
        filter.addAction(MusicPlaybackService.META_CHANGED);
        filter.addAction(MusicPlaybackService.PREPARED);
        filter.addAction(MusicPlaybackService.REFRESH);
        requireActivity().registerReceiver(mPlaybackStatus, filter);
    }

    @Override
    public boolean onBackPressed() {
        return true;
    }

    @Override
    public void onPause() {
        try {
            requireActivity().unregisterReceiver(mPlaybackStatus);
        } catch (final Throwable ignored) {
        }
        super.onPause();
    }

    private final class PlaybackStatus extends BroadcastReceiver {
        @Override
        public void onReceive(final Context context, final Intent intent) {
            final String action = intent.getAction();
            if (isNull(action)) return;

            switch (action) {
                case MusicPlaybackService.PLAYSTATE_CHANGED:
                    mAdapter.notifyDataSetChanged();
                    break;
            }
        }
    }
}

File: app/src/main/java/biz/dealnote/messenger/fragment/PlaylistFragment.kt
package biz.dealnote.messenger.fragment

import android.app.Dialog
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.ItemTouchHelper
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import biz.dealnote.messenger.Extra
import biz.dealnote.messenger.R
import biz.dealnote.messenger.adapter.AudioRecyclerAdapter
import biz.dealnote.messenger.listener.BackPressCallback
import biz.dealnote.messenger.model.Audio
import biz.dealnote.messenger.place.PlaceFactory
import biz.dealnote.messenger.player.MusicPlaybackService
import biz.dealnote.messenger.player.MusicPlaybackService.Companion.startForPlayList
import biz.dealnote.messenger.player.util.MusicUtils
import biz.dealnote.messenger.settings.Settings
import biz.dealnote.messenger.util.Objects
import biz.dealnote.messenger.util.PhoenixToast.Companion.CreatePhoenixToast
import biz.dealnote.messenger.util.Utils
import com.google.android.material.bottomsheet.BottomSheetBehavior
import com.google.android.material.bottomsheet.BottomSheetDialog
import com.google.android.material.bottomsheet.BottomSheetDialogFragment
import com.google.android.material.floatingactionbutton.FloatingActionButton
import java.util.*

class PlaylistFragment : BottomSheetDialogFragment(), AudioRecyclerAdapter.ClickListener, BackPressCallback {
    private var mRecyclerView: RecyclerView? = null
    private var mAdapter: AudioRecyclerAdapter? = null
    private var mData: ArrayList<Audio>? = null
    private var mPlaybackStatus: PlaybackStatus? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        mData = requireArguments().getParcelableArrayList(Extra.AUDIOS)
    }

    private fun getAudioPos(audio: Audio): Int {
        if (mData != null && mData!!.isNotEmpty()) {
            for ((pos, i) in mData!!.withIndex()) {
                if (i.id == audio.id && i.ownerId == audio.ownerId) {
                    i.isAnimationNow = true
                    mAdapter!!.notifyDataSetChanged()
                    return pos
                }
            }
        }
        return -1
    }

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        val dialog = BottomSheetDialog(requireActivity(), theme)
        val behavior = dialog.behavior
        behavior.state = BottomSheetBehavior.STATE_EXPANDED
        behavior.skipCollapsed = true
        return dialog
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val root = inflater.inflate(R.layout.fragment_playlist, container, false)
        mRecyclerView = root.findViewById(R.id.list)
        val manager = LinearLayoutManager(requireActivity(), RecyclerView.VERTICAL, false)
        mRecyclerView?.layoutManager = manager
        val Goto: FloatingActionButton = root.findViewById(R.id.goto_button)
        Goto.setOnLongClickListener {
            val curr = MusicUtils.getCurrentAudio()
            if (curr != null) {
                PlaceFactory.getPlayerPlace(Settings.get().accounts().current).tryOpenWith(requireActivity())
            } else CreatePhoenixToast(requireActivity()).showToastError(R.string.null_audio)
            false
        }
        Goto.setOnClickListener {
            val curr = MusicUtils.getCurrentAudio()
            if (curr != null) {
                val index = getAudioPos(curr)
                if (index >= 0) {
                    if (Settings.get().other().isShow_audio_cover) mRecyclerView?.scrollToPosition(index) else mRecyclerView?.smoothScrollToPosition(index)
                } else CreatePhoenixToast(requireActivity()).showToast(R.string.audio_not_found)
            } else CreatePhoenixToast(requireActivity()).showToastError(R.string.null_audio)
        }
        ItemTouchHelper(simpleItemTouchCallback).attachToRecyclerView(mRecyclerView)
        return root
    }

    var simpleItemTouchCallback: ItemTouchHelper.SimpleCallback = object : ItemTouchHelper.SimpleCallback(0, ItemTouchHelper.LEFT) {
        override fun onMove(recyclerView: RecyclerView,
                            viewHolder: RecyclerView.ViewHolder, target: RecyclerView.ViewHolder): Boolean {
            return false
        }

        override fun onSwiped(viewHolder: RecyclerView.ViewHolder, swipeDir: Int) {
            Utils.vibrate(requireActivity(), 100)
            mAdapter?.notifyDataSetChanged()
            startForPlayList(requireActivity(), mData!!, mAdapter!!.getItemRawPosition(viewHolder.layoutPosition), false)
        }
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        mAdapter = AudioRecyclerAdapter(requireActivity(), mData, false, false, 0)
        mAdapter!!.setClickListener(this)
        mRecyclerView!!.adapter = mAdapter
        val my = MusicUtils.getCurrentAudio()
        if (my != null) {
            var index = 0
            var o = 0
            for (i in mData!!) {
                if (i === my) {
                    index = o
                    break
                }
                o++
            }
            mRecyclerView!!.scrollToPosition(index)
        }
    }

    override fun onClick(position: Int, catalog: Int, audio: Audio) {
        startForPlayList(requireActivity(), mData!!, position, false)
    }

    override fun onResume() {
        super.onResume()
        mPlaybackStatus = PlaybackStatus()
        val filter = IntentFilter()
        filter.addAction(MusicPlaybackService.PLAYSTATE_CHANGED)
        filter.addAction(MusicPlaybackService.SHUFFLEMODE_CHANGED)
        filter.addAction(MusicPlaybackService.REPEATMODE_CHANGED)
        filter.addAction(MusicPlaybackService.META_CHANGED)
        filter.addAction(MusicPlaybackService.PREPARED)
        filter.addAction(MusicPlaybackService.REFRESH)
        requireActivity().registerReceiver(mPlaybackStatus, filter)
    }

    override fun onBackPressed(): Boolean {
        return true
    }

    override fun onPause() {
        try {
            requireActivity().unregisterReceiver(mPlaybackStatus)
        } catch (ignored: Throwable) {
        }
        super.onPause()
    }

    private inner class PlaybackStatus : BroadcastReceiver() {
        override fun onReceive(context: Context, intent: Intent) {
            val action = intent.action
            if (Objects.isNull(action)) return
            if (MusicPlaybackService.PLAYSTATE_CHANGED == action) {
                mAdapter!!.notifyDataSetChanged()
            }
        }
    }

    companion object {
        fun buildArgs(playlist: ArrayList<Audio?>?): Bundle {
            val bundle = Bundle()
            bundle.putParcelableArrayList(Extra.AUDIOS, playlist)
            return bundle
        }

        fun newInstance(playlist: ArrayList<Audio?>?): PlaylistFragment {
            return newInstance(buildArgs(playlist))
        }

        fun newInstance(args: Bundle?): PlaylistFragment {
            val fragment = PlaylistFragment()
            fragment.arguments = args
            return fragment
        }
    }
}